File: Code/CH8 Code/VB/CustomValidator-vb/Validator.vb
Option Strict On

Imports System.ComponentModel
Imports System.Windows.Forms

<ProvideProperty("Required", GetType(Control))> _
Public Class Validator
  Inherits Component
  Implements IExtenderProvider

  Private Props As Hashtable = New Hashtable()


  Function CanExtend(ByVal target As Object) As Boolean Implements IExtenderProvider.CanExtend
    If TypeOf target Is TextBox Or TypeOf target Is ComboBox Then
      Return True
    Else
      Return False
    End If
  End Function

  Function GetRequired(ByVal ctl As Control) As Boolean
    If Props.Contains(ctl) Then
      Return CBool(Props.Item(ctl))
    Else
      Return False
    End If
  End Function


End Class

File: Code/CH8 Code/C#/CutomValidator-c/Validator.cs
using System;
using System.Collections;
using System.Windows.Forms;
using System.ComponentModel;

namespace CutomValidator_c
{

  [ProvideProperty("Required", typeof(Control))]
  public class Validator : Component,
                           IExtenderProvider
	{

    private System.Collections.Hashtable Props = new System.Collections.Hashtable();

		public Validator()
		{
    }

    bool IExtenderProvider.CanExtend(object target) 
    {
      if (target is TextBox || target is ComboBox) 
        return true;
      else
        return false;
    }


    public bool GetRequired(Control ctl)
    {
      if(Props.Contains(ctl))
        return (bool)Props[ctl];
      else
        return false;
    }

    public void SetRequired(Control ctl, bool val)
    {
      if(Props.Contains(ctl))
        Props[ctl] = val;
      else
      {
        Props.Add(ctl, val);
        ctl.Validating += new CancelEventHandler(ValidateProp);
      }
    }

    private void ValidateProp(object sender, CancelEventArgs e)
    {
      Control ctl = (Control)sender;
      if(GetRequired(ctl))
        if(ctl.Text == string.Empty)
          e.Cancel = true;
    }

  }
}
